pyteal/ast/abi/tuple.py

```diff
@@ -484,3 +486,57 @@ def __init__(
 
 
 Tuple5.__module__ = "pyteal.abi"
+
+
+class NamedTupleTypeSpec(TupleTypeSpec):
+    """A NamedTupleType inherits from TupleTypeSpec, allowing for more than 5 elements."""
+
+    def __init__(
+        self, instance_class: type["NamedTuple"], *value_type_specs: TypeSpec
+    ) -> None:
+        if instance_class == NamedTuple:
+            raise TealInputError(
+                "NamedTupleTypeSpec must be instanced with subclassed NamedTuple class."
+            )
+
+        self.instance_class: type["NamedTuple"] = instance_class
+        super().__init__(*value_type_specs)
+
+    def annotation_type(self) -> "type[NamedTuple]":
+        return self.instance_class
+
+    def new_instance(self) -> "NamedTuple":
+        return self.instance_class()
+
+
+NamedTupleTypeSpec.__module__ = "pyteal.abi"
+
+
+class NamedTuple(Tuple):
+    """A NamedTuple is a Tuple with all its elements named."""
+
+    def __init__(self):
+        if type(self) is NamedTuple:
+            raise TealInputError("NamedTuple must be subclassed.")
+
+        anns = get_annotations(type(self))
+        if not anns:
+            raise Exception("Expected fields to be declared but found none")
+
+        self.type_specs: OrderedDict[str, TypeSpec] = OrderedDict()
+
+        for index, (name, annotation) in enumerate(anns.items()):
+            self.type_specs[name] = type_spec_from_annotation(annotation)
+
+        super().__init__(
+            NamedTupleTypeSpec(type(self), *list(self.type_specs.values()))
+        )
+
+        for index, name in enumerate(anns):
+            setattr(self, name, self[index])
```

Comment:
One potential issue with this implementation is that it doesn't prevent users from assigning new values to these fields. For example:

```python
instance = NamedTupleInherit()
instance.c = True

# now instance.c returns True instead of indexing into the instance value
program = If(instance.c).Then(...)
```

If this is something we care to fix, one solution would be to override `__getattr__` and `__setattr__` for this class. I attempted to do this locally and I think you can have success with this approach:

```diff
diff --git a/pyteal/ast/abi/tuple.py b/pyteal/ast/abi/tuple.py
index 5da00d9..7f7f8df 100644
--- a/pyteal/ast/abi/tuple.py
+++ b/pyteal/ast/abi/tuple.py
@@ -523,17 +523,29 @@ class NamedTuple(Tuple):
         if not anns:
             raise Exception("Expected fields to be declared but found none")
 
+        self._ready = False
+
         self.type_specs: OrderedDict[str, TypeSpec] = OrderedDict()
+        self.indexes: dict[str, int] = dict()
 
         for index, (name, annotation) in enumerate(anns.items()):
             self.type_specs[name] = type_spec_from_annotation(annotation)
+            self.indexes[name] = index
 
         super().__init__(
             NamedTupleTypeSpec(type(self), *list(self.type_specs.values()))
         )
 
-        for index, name in enumerate(anns):
-            setattr(self, name, self[index])
+        self._ready = True
+    
+    def __getattr__(self, name: str) -> TupleElement[Any]:
+        return self.__getitem__(self.indexes[name])
+    
+    def __setattr__(self, name: str, value: Any) -> None:
+        if name == "_ready" or not self._ready:
+            super().__setattr__(name, value)
+            return
+        raise TealInputError("Cannot assign to NamedTuple attributes")
 
     def type_spec(self) -> TupleTypeSpec:
         return cast(NamedTupleTypeSpec, super().type_spec())

```

Reply:
good point, but I would like to know what is the `self._ready` for?

Reply:
oh snap I kind of see it coming. I suppose it is for avoiding erroring in `NamedTuple` level assigning, but still error on attribute field level?

Reply:
The `__setattr__` method is invoked when the attributes `type_specs`, `indexes`, and all attributes from `super().__init__()` are set. In order to allow these attributes to be set like normal, I introduced the `_ready` variable, which essentially disables the `__setattr__` method until after the `__init__` method has finished.

There may be others ways to achieve the same thing, but this is what I came up with.

Reply:
really good point, I think it involves a few things here in Python class initialization, and class hidden member naming convention, for I tried `__ready` and it does not quite work.

Essentially we are disabling `__setattr__` after `__init__` is done for executing. I will add some notes to explain what is going on. Thanks for explaining.